PowerEditor/src/winmain.cpp

```diff
@@ -548,51 +554,126 @@ int WINAPI wWinMain(HINSTANCE hInstance, HINSTANCE, PWSTR pCmdLine, int)
+		};
+
+		// If we are set to be environment-aware, initialize the COM subsystem, and try to get an interface pointer to the VDM
+		if (nppGui._environmentAware && SUCCEEDED(::CoInitialize(nullptr)))
 		{
-			Sleep(100);
-			hNotepad_plus = ::FindWindow(Notepad_plus_Window::getClassName(), NULL);
-		}
+			IVirtualDesktopManager* pVDM = nullptr;
+			if (SUCCEEDED(::CoCreateInstance(CLSID_VirtualDesktopManager, nullptr, CLSCTX_ALL, IID_PPV_ARGS(&pVDM))))
+			{
+				// We can only pass a single LPARAM to the EnumWindows callback, so we will pack both the VDM pointer
+				// and the result boolean into a single pair, and pass that.
+				std::pair<IVirtualDesktopManager*, bool> enumWindowsDataPair = { pVDM, false };
+
+				static const auto TestWindow = [](IVirtualDesktopManager* const pVDM, HWND hWnd) -> bool {
+					// Compare the class name of the window against our expected class name.
+					// We only need to compare one additional character to know if it's different.
+					TCHAR classNameBuffer[_countof(Notepad_plus_Window::ClassName) + 1] = { _T('\0') };
+					if (SUCCEEDED(::GetClassName(hWnd, classNameBuffer, _countof(classNameBuffer))))
+					{
+						if (generic_strncmp(Notepad_plus_Window::ClassName, classNameBuffer, _countof(classNameBuffer)) == 0)
+						{
+							// This window is another Notepad++ window. Is it in our virtual desktop?
+							BOOL onCurrentVirtualDesktop = TRUE;
+							if (SUCCEEDED(pVDM->IsWindowOnCurrentVirtualDesktop(hWnd, &onCurrentVirtualDesktop)) && onCurrentVirtualDesktop)
+							{
+								// This is a Notepad++ window that is on our current virtual desktop.
+								return true;
+							}
+						}
+					}
+
+					return false;
+				};
 
-        if (hNotepad_plus)
-        {
-			// First of all, destroy static object NppParameters
-			nppParameters.destroyInstance();
+				::EnumWindows([](HWND hWnd, LPARAM lParam) -> BOOL
+				{
+					auto& inOutData = *reinterpret_cast<decltype(enumWindowsDataPair)*>(lParam);
+
+					// If we find a window that is a Notepad++ window that is a part of our current virtual desktop,
+					// report it back to the result pair and stop enumeration.
+					if (TestWindow(inOutData.first, hWnd)) {
+						inOutData.second = true;
+						return FALSE;
+					}
 
-			int sw = 0;
+					return TRUE;
+				}, reinterpret_cast<LPARAM>(&enumWindowsDataPair));
+				
+				// If the pair's second parameter is 'true', then an existing window was found on this virtual desktop.
+				if (!enumWindowsDataPair.second)
+				{
+					// Otherwise, let's just do a normal, quick test using FindWindow just as a sanity-check
+					if (HWND appWindow = FindAppWindow())
+					{
+						enumWindowsDataPair.second = TestWindow(pVDM, appWindow);
+					}
 
-			if (::IsZoomed(hNotepad_plus))
-				sw = SW_MAXIMIZE;
-			else if (::IsIconic(hNotepad_plus))
-				sw = SW_RESTORE;
+					shouldTerminate = enumWindowsDataPair.second;
+				}
 
-			if (sw != 0)
-				::ShowWindow(hNotepad_plus, sw);
+				pVDM->Release();

```

Comment:
Minor suggestion: You can use [com_ptr](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp#L40) or similar smart pointer to avoid memory leaks.
If you chose the former, then its declaration should probably be moved to some header.

Reply:
I had actually implemented a rather deep `ComPtr` type that included reference counting and awareness of the thread that it was initialized in (due to `CoInitialize` semantics) before realizing that `com_ptr` existed, but had removed it to keep the PR smaller :|.